File: qt/userifc_py/qt/resources/qt/helloqtquickdemo.py
#!/usr/bin/env python -t

from __future__ import (absolute_import, division, print_function,
  unicode_literals)

import sys, os, logging, inspect
from future.builtins import (ascii, filter, hex, map, oct, zip, str)

os.environ['QT_QUICK_BACKEND'] = os.environ.get('QT_QUICK_BACKEND', 'software')

if '4' != os.environ.get('QT_MAJOR_VERSION', 'LATEST'):
  from PySide2 import QtCore, QtGui, QtWidgets, QtQml
  Slot = QtCore.Slot
  #from PyQt5 import sip, Qt, QtCore, QtWidgets, QtQml
  #Slot = QtCore.pyqtSlot
  #sip.setapi('QVariant', 2)
  QApplication = QtWidgets.QApplication
  _uiform = 'qt/helloForm-qt5.qml'
else:
  from PySide import QtCore, QtGui, QtQml
  Slot = QtCore.Slot
  #from PyQt4 import Qt, QtCore, QtGui, QtDeclarative
  #Slot = QtCore.pyqtSlot
  #import sip
  #sip.setapi('QVariant', 2)
  QApplication = QtGui.QApplication
  _uiform = 'qt/helloForm-qt4.qml'

def userifc_version():
  import platform

  return "(Python {0}) Qt {1}".format(platform.python_version(),
    QtCore.qVersion())
  #return "(Python {0}) Qt {1}".format(platform.python_version(),
  #  Qt.QT_VERSION_STR)

class Handler(QtCore.QObject):
  def __init__(self, parent):
    super(QtCore.QObject, self).__init__()
    self.view1 = parent
  
  @Slot()
  def on_button1_clicked(self):
    '''Callback on button1 clicked'''
    
    self.view1.widgets['simplebutton1'].setProperty('opacity', 0)
    self.view1.widgets['entry1'].setProperty('text', '')
    self.view1.widgets['simpleentry1'].setProperty('opacity', 1)
    self.view1.widgets['entry1'].setProperty('focus', 1)
  
  @Slot()
  def on_entry1_returnPressed(self):
    '''Callback on entry1 returnPressed'''
    
    self.view1.widgets['simpleentry1'].setProperty('opacity', 0)
    self.view1.widgets['simplebutton1'].setProperty('opacity', 1)
    self.view1.widgets['textview1'].setProperty('text', 'Hello, ' +
      self.view1.widgets['entry1'].property('text') + '.')

class HelloQtquickDemo(QtCore.QObject):
  '''Description:: '''

  def __init__(self, parent=None):
    '''Construct object'''
    
    # QtCore.QObject.__init__(self, parent)
    super(HelloQtquickDemo, self).__init__(parent)
    
    self.widgets = {}
    rsrc_path = os.environ.get('RSRC_PATH', 'resources')
    if '4' != os.environ.get('QT_MAJOR_VERSION', 'LATEST'):
      self.engine = QtQml.QQmlApplicationEngine(rsrc_path + '/' + _uiform)
      #self.engine.rootContext().setContextProperty('handler', Hdlr)
      self.rootObject = self.engine.rootObjects()[0]
      self.decl_view = self.rootObject
    else:
      self.decl_view = QtDeclarative.QDeclarativeView(
        QtCore.QUrl.fromLocalFile(rsrc_path + '/' + _uiform))
      self.decl_view.setResizeMode(
        QtDeclarative.QDeclarativeView.SizeRootObjectToView)
      #self.engine.rootContext().setContextProperty('handler', Hdlr)
      self.rootObject = self.decl_view.rootObject()
      self.engine = self.decl_view.engine()
    
    #[self.button1, self.entry1, self.textview1, self.simplebutton1,
    #  self.simpleentry1] = map(lambda nm: self.rootObject.findChild(
    #  QtCore.QObject, nm), ['button1', 'entry1', 'textview1',
    #  'simplebutton1', 'simpleentry1'])
    for name in ['button1', 'entry1', 'textview1', 'simplebutton1',
        'simpleentry1']:
      self.widgets[name] = self.rootObject.findChild(QtCore.QObject, name)
    
    #self.hdlr = Handler(self)
    #self.engine.rootContext().setContextProperty('handler', self.hdlr)
    self.engine.rootContext().setContextProperty('handler', self)
    self.decl_view.setGeometry(400, 300, 240, 160)
    self.decl_view.show()
  
  @Slot()
  def on_button1_clicked(self):
    '''Callback on button1 clicked'''
    
    self.widgets['simplebutton1'].setProperty('opacity', 0)
    self.widgets['entry1'].setProperty('text', '')
    self.widgets['simpleentry1'].setProperty('opacity', 1)
    self.widgets['entry1'].setProperty('focus', 1)
  
  @Slot()
  def on_entry1_returnPressed(self):
    '''Callback on entry1 returnPressed'''
    
    self.widgets['simpleentry1'].setProperty('opacity', 0)
    self.widgets['simplebutton1'].setProperty('opacity', 1)
    self.widgets['textview1'].setProperty('text', 'Hello, ' +
      self.widgets['entry1'].property('text') + '.')

def main(argv=None):
  app = QApplication([])
  pretext = '{0} GUI\n'.format(userifc_version())
  gui = HelloQtquickDemo()
  gui.widgets['textview1'].setProperty('text', pretext)
  
  app.exec_()
  return 0

if '__main__' == __name__:
  sys.exit(main(sys.argv[1:]))


# (one-time): pip install --user pyside2 pyqt5
# [QT_QUICK_BACKEND=software] python helloqtquickdemo.py

File: qt/userifc_py/qt/resources/qt/HelloQtquickDemo.cpp
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <string>
#include <QtGui>
#include <QtQml>
#include <QQmlApplicationEngine>
//#include <QDeclarativeView>
//#include <QDeclarativeEngine>
#include <QApplication>

//#define QStringLiteral(str) QString::fromUtf8("" str "", sizeof(str) - 1)

class HelloQtquickDemo : public QObject {
    Q_OBJECT

public:
    explicit HelloQtquickDemo(QObject *parent = 0);
    
    //virtual ~HelloQtquickDemo(void);
    
    QQmlApplicationEngine *engine;
    QObject *rootObject;
    QWindow *decl_view;
    
    QObject *button1;
    QObject *entry1;
    QObject *textview1;
    QObject *simplebutton1;
    QObject *simpleentry1;

public slots:
    void on_button1_clicked(void);
    void on_entry1_textChanged(QString);
    void on_entry1_returnPressed(void);

private:
    
};

HelloQtquickDemo::HelloQtquickDemo(QObject *parent) : QObject(parent) {
    std::string rsrc_path = getenv("RSRC_PATH") ? getenv("RSRC_PATH") : 
        "resources";
    //QT_VERSION_MAJOR depends on compile: Qt5Core or Qt4Core
    std::string uiform = rsrc_path + "/" + ((4 == QT_VERSION_MAJOR) ? 
        "qt/helloForm-qt4.qml" : "qt/helloForm-qt5.qml");
    
    if (4 != QT_VERSION_MAJOR) {
        //this->engine = new QQmlApplicationEngine(uiform.c_str());
        //this->engine = new QQmlApplicationEngine(
        //    QUrl::fromLocalFile(QStringLiteral(uiform)));
        this->engine = new QQmlApplicationEngine();
        //this->engine->load(QUrl::fromLocalFile(
        //    QStringLiteral(uiform.c_str())));
        this->engine->load(QUrl::fromLocalFile(uiform.c_str()));
        this->rootObject = engine->rootObjects().first();
        this->decl_view = (QWindow*)this->rootObject;
    } else {
        /*//this->decl_view = new QDeclarativeView(uiform.c_str());
        this->decl_view = new QDeclarativeView(
            QUrl::fromLocalFile(QStringLiteral(uiform)));
        this->decl_view->setResizeMode(
            QDeclarativeView::SizeRootObjectToView);
        this->rootObject = this->decl_view->rootObject();
        this->engine = this->decl_view->engine();*/
    }
    
    this->button1 = this->rootObject->findChild<QObject*>("button1");
    this->entry1 = this->rootObject->findChild<QObject*>("entry1");
    this->textview1 = this->rootObject->findChild<QObject*>("textview1");
    this->simplebutton1 = this->rootObject->findChild<QObject*>(
        "simplebutton1");
    this->simpleentry1 = this->rootObject->findChild<QObject*>(
        "simpleentry1");
    
    /*QObject::connect(button1, SIGNAL(clicked()), this, 
        SLOT(on_button1_clicked()));
    QObject::connect(entry1, SIGNAL(textChanged(QString)), this, 
        SLOT(on_entry1_textChanged(QString)));
    QObject::connect(entry1, SIGNAL(returnPressed()), this, 
        SLOT(on_entry1_returnPressed()));*/
    this->engine->rootContext()->setContextProperty("handler", this);
    
    this->decl_view->setGeometry(400, 300, 240, 160);
    this->decl_view->show();
}

//HelloQtquickDemo::~HelloQtquickDemo(void) {
//    ;
//}

void HelloQtquickDemo::on_button1_clicked(void) {
    this->simplebutton1->setProperty("opacity", 0);
    this->entry1->setProperty("text", "");
    this->simpleentry1->setProperty("opacity", 1);
    this->entry1->setProperty("focus", 1);
}

void HelloQtquickDemo::on_entry1_textChanged(QString) {
    this->textview1->setProperty("text", "Hello, " + 
        this->entry1->property("text").toString() + ".");
}

void HelloQtquickDemo::on_entry1_returnPressed(void) {
    this->simpleentry1->setProperty("opacity", 0);
    this->simplebutton1->setProperty("opacity", 1);
    this->textview1->setProperty("text", "Hello, " + 
        this->entry1->property("text").toString() + ".");
}

int main(int argc, char *argv[]) {
    //fprintf(stderr, "QT_VERSION_MAJOR: %d\n", QT_VERSION_MAJOR);
    setenv("QT_QUICK_BACKEND", getenv("QT_QUICK_BACKEND") ?
        getenv("QT_QUICK_BACKEND") : "software", 1);
    QApplication app(argc, argv);
    
    char pretextBuf[256];
    snprintf(pretextBuf, sizeof(pretextBuf) - 1,
        "(GCC %d.%d) Qt version: %s\n", __GNUC__, 
        __GNUC_MINOR__, qVersion());

    HelloQtquickDemo gui;
    //QObject::connect(gui.engine, SIGNAL(quit()), &app, SLOT(quit()));
    gui.textview1->setProperty("text", pretextBuf);

    exit(app.exec());
}

#include "moc_HelloQtquickDemo.cpp"


// moc HelloQtquickDemo.cpp -o build/moc_HelloQtquickDemo.cpp
// c++ -fPIC -I build HelloQtquickDemo.cpp -o build/main-HelloQtquickDemo `pkg-config --cflags --libs QtCore QtGui QtWidgets QtQml opengl`
// build/main-HelloQtquickDemo
